src/app/base/side-panel-context-new.tsx

```diff
@@ -0,0 +1,99 @@
+import type { ComponentType, PropsWithChildren, ReactElement } from "react";
+import { useContext, createContext, useState } from "react";
+
+// This new side panel context is meant to replace the old one over time. Since migrating all
+// panels will take some time, these two contexts will live together while the migrations are
+// being carried out.
+// TODO: Once the migrations are complete, remove the old side panel context.
+
+type SidePanelSize = "large" | "narrow" | "regular" | "wide";
+
+interface SidePanelState<TProps = Record<string, unknown>> {
+  isOpen: boolean;
+  title: string;
+  size: SidePanelSize;
+  component: ComponentType<TProps> | null;
+  props: TProps;
+}
+
+interface SidePanelActions {
+  open: <TProps extends Record<string, unknown> = Record<string, unknown>>(
+    component: ComponentType<TProps>,
+    title: string,
+    props?: TProps,
+    size?: SidePanelSize
+  ) => void;
+  close: () => void;
+  setSize: (size: SidePanelSize) => void;
+}
+
+type SidePanelContextValue = SidePanelActions & SidePanelState;
+
+const SidePanelContext = createContext<SidePanelContextValue | null>(null);
+
+export const useSidePanel = (): SidePanelContextValue => {
+  const context = useContext(SidePanelContext);
+  if (!context) {
+    throw new Error("useSidePanel must be used within a SidePanelProvider");
+  }
+  return context;
+};
+
+const SidePanelContextProvider = ({
+  children,
+}: PropsWithChildren): ReactElement => {
+  const [state, setState] = useState<SidePanelState>({
+    isOpen: false,
+    title: "",
+    size: "regular",
+    component: null,
+    props: {},
+  });
+
+  const open = <
+    TProps extends Record<string, unknown> = Record<string, unknown>,
+  >(
+    component: ComponentType<TProps>,
+    title: string,
+    props: TProps = {} as TProps,
+    size: SidePanelSize = "regular"
+  ) => {
```

Comment:
We should use named parameters if a function has more than 2 parameters

```suggestion
  const open = <
    TProps extends Record<string, unknown> = Record<string, unknown>,
  >({ component, title, props = {} as TProps, size = "regular" }: {
    component: ComponentType<TProps>;
    title: string;
    props: TProps;
    size: SidePanelSize;
  }) => {
```